### PR TITLE
feat: CP support Menu expandIcon

### PR DESCRIPTION
**Pull request by li-jia-nan**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 💡 Background and solution

- #46889

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | feat: CP support Menu expandIcon |
| 🇨🇳 Chinese | feat: CP support Menu expandIcon |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/menu/menu.tsx [hunk 1]
```diff
@@ -126,22 +127,21 @@ const InternalMenu = forwardRef<RcMenuRef, InternalMenuProps>((props, ref) => {
...
+    }
+    return cloneElement(waitClone, {
       className: classNames(
         `${prefixCls}-submenu-expand-icon`,
-        isValidElement(beClone) ? beClone.props?.className : '',
+        React.isValidElement(waitClone) ? waitClone.props?.className : undefined,
       ),
     });
-  }
+  }, [expandIcon, overrideObj.expandIcon, menu?.expandIcon]);
```
Comment (MadCcc) [inline components/menu/menu.tsx:144]: 这里感觉 useMemo 有点多余，只是一些简单的 if else

File: components/menu/menu.tsx [hunk 2]
```diff
@@ -126,22 +127,21 @@ const InternalMenu = forwardRef<RcMenuRef, InternalMenuProps>((props, ref) => {
...
+      return waitClone;
+    }
+    if (waitClone === null || waitClone === false) {
+      return null;
+    }
+    return cloneElement(waitClone, {
       className: classNames(
         `${prefixCls}-submenu-expand-icon`,
-        isValidElement(beClone) ? beClone.props?.className : '',
+        React.isValidElement(waitClone) ? waitClone.props?.className : undefined,
```
Comment (MadCcc) [inline components/menu/menu.tsx:141]: 这里原本不是用的 React 的 isValidElement

Comment (MadCcc) [reply components/menu/menu.tsx:141]: waitClone 这命名好怪

File: components/menu/menu.tsx [hunk 3]
```diff
@@ -126,22 +127,21 @@ const InternalMenu = forwardRef<RcMenuRef, InternalMenuProps>((props, ref) => {
...
-    mergedExpandIcon = expandIcon;
-  } else if (expandIcon === null || expandIcon === false) {
-    mergedExpandIcon = null;
-  } else if (overrideObj.expandIcon === null || overrideObj.expandIcon === false) {
-    mergedExpandIcon = null;
-  } else {
-    const beClone: React.ReactNode = (expandIcon ?? overrideObj.expandIcon) as React.ReactNode;
-    mergedExpandIcon = cloneElement(beClone, {
+  const mergedExpandIcon = React.useMemo<RenderIconType>(() => {
+    const waitClone = expandIcon ?? overrideObj.expandIcon ?? menu?.expandIcon;
```
Comment (MadCcc) [inline components/menu/menu.tsx:131]: 这里逻辑不太一致了，null 在原本的逻辑里是特判的

Comment (li-jia-nan) [reply components/menu/menu.tsx:131]: 是的，CI 已经挂了，我改一下

File: components/menu/menu.tsx (hunk 2 above)
Comment (li-jia-nan) [reply components/menu/menu.tsx]: 改成 mergedIcon 了

File: components/menu/menu.tsx (hunk 1 above)
Comment (li-jia-nan) [reply components/menu/menu.tsx]: 改回去了

File: components/menu/menu.tsx [hunk 4]
```diff
@@ -126,19 +127,23 @@ const InternalMenu = forwardRef<RcMenuRef, InternalMenuProps>((props, ref) => {
   const menuClassName = classNames(`${prefixCls}-${theme}`, menu?.className, className);
 
   // ====================== Expand Icon ========================
-  let mergedExpandIcon: MenuProps['expandIcon'];
+  let mergedExpandIcon: RenderIconType;
```
Comment (MadCcc) [inline components/menu/menu.tsx:130]: 这里用 `MenuProps['expandIcon']` 更好吧

Comment (li-jia-nan) [reply components/menu/menu.tsx]: done

File: components/menu/menu.tsx [hunk 5]
```diff
@@ -126,19 +127,23 @@ const InternalMenu = forwardRef<RcMenuRef, InternalMenuProps>((props, ref) => {
...
   if (typeof expandIcon === 'function') {
     mergedExpandIcon = expandIcon;
   } else if (expandIcon === null || expandIcon === false) {
     mergedExpandIcon = null;
   } else if (overrideObj.expandIcon === null || overrideObj.expandIcon === false) {
     mergedExpandIcon = null;
+  } else if (typeof menu?.expandIcon === 'function') {
+    mergedExpandIcon = menu?.expandIcon;
+  } else if (menu?.expandIcon === null || menu?.expandIcon === false) {
+    mergedExpandIcon = null;
```
Comment (MadCcc) [inline components/menu/menu.tsx:139]:
```
const mergeExpandIcon = getValidIcon(expandIcon, overrideObj.expandIcon, menu?.expandIcon);

function getValidIcon(...args: MenuProps['expandIcon'][]) {
  for (const icon of args) {
    if (typeof icon === 'function') {
      return icon;
    } else if (icon === null || icon === false) {
      return null;
    }
  }
}
```
一点想法，可以试试，可能还有更好的写法

Comment (li-jia-nan) [reply components/menu/menu.tsx]: 最后还是用 `useMemo` 了，就三个 case，感觉没有必要用 for 循环呢

**Review by MadCcc — APPROVED** [review]

**Pull request merged**